src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp

```diff
@@ -42,47 +42,60 @@ inline T UnifiedOopRef::addr() const {
 // this specialization provides a workaround.
 template<>
 inline uintptr_t UnifiedOopRef::addr<uintptr_t>() const {
-  return _value & ~uintptr_t(3);
+  return (_value & ~tag_mask) LP64_ONLY(>> 1);
```

Comment:
Why a shift now as well?

Reply:
I do not remember exactly. But I think there was some ref which on 64 bit which had the third bit set. That is this assert failed:
`assert((reinterpret_cast<uintptr_t>(ref) & UnifiedOopRef::tag_mask) == 0, "Unexpected low-order bits");`
I have to remove the shift and test again. I think it is strange given that `ObjectAlignmentInBytes >= 8` but we added the shift for some reason. Or maybe I am just missing something obvious.

Reply:
ObjectAlignmentInBytes is 4 by default on 32 bit VMs.

Reply:
@fisk Why is the shifts needed on 64-bits?

Reply:
Maybe the shift is needed in order to handle unaligned oops, as for code blob oops? The existing root processing logic has skipped code blob oops, mainly because they were hard to encode. The rationale was that they are not roots in an of themselves. Now, we have seen several instances where chains are not found on iterations, so maybe the assumption on skipping code blob oops is wrong? With this encoding scheme in place, it would be possible to pass a closure for processing the code blob oops as well.

Reply:
Sorry for the tardy response. 

I think I forgot what the code was actually doing at the time of writing my previous answer. We are encoding ` narrowOop*` and `oop*` here not `narrowOop`/`oop`. And `narrowOop*` are 4 byte aligned.

Reply:
Ok.